### PR TITLE
[flang][OpenMP] Fix 2 more regressions after #101009

### PR DESCRIPTION
**Pull request by luporl**
PR #101009 exposed a semantic check issue with OPTIONAL dummy
arguments.
Another issue occurred when using %{re,im,len,kind}, as these also
need to be skipped when handling variables with implicitly defined
DSAs.

These issues were found by Fujitsu testsuite.

### CODE REVIEW

**Comment by tblah** [comment]
Our tests still pass after this change. Please wait for approval from @psteinfeld

**Comment by luporl** [comment]
ping

**Review by psteinfeld — APPROVED** [review]
All builds and tests correctly and looks good.

**Comment by luporl** [comment]
Thanks for reviewing and testing this PR @tblah and @psteinfeld.

**Pull request merged**
